test/store.js

```diff
@@ -13,31 +19,44 @@ describe('Store: inmemory.js', () => {
     store = new Store();
   });
 
-  it('lists all keys', () => {
-    store.put(key1, value1);
-    store.put(key2, value2);
-
-    expect(store.keys()).to.contain(key1, key2);
-  });
+  describe('read and write is available without any locks', () => {
+    it('gets a key', () => {
+      store.put(key1, value1);
 
-  it('gets a key', () => {
-    store.put(key1, value1);
-
-    expect(store.get(key1)).to.equal(value1);
-  });
+      expect(store.get(key1)).to.equal(value1);
+    });
 
-  it('deletes a key', () => {
-    store.put(key1, value1);
-    store.delete(key1);
-    // eslint-disable-next-line no-unused-expressions
-    expect(store.get(key1)).to.be.null;
+    it('clears the store', () => {
+      store.put(key1, value1);
+      store.put(key2, value2);
+      store.clear();
+      // eslint-disable-next-line no-unused-expressions
+      expect(store.keys()).to.be.empty;
+    });
   });
 
-  it('clears the store', () => {
-    store.put(key1, value1);
-    store.put(key2, value2);
-    store.clear();
-    // eslint-disable-next-line no-unused-expressions
-    expect(store.keys()).to.be.empty;
+  describe('lock', () => {
+    afterEach(() => store.clear());
+
+    it('prevents race condition', () =>
+      expect(Promise.all([timeoutLock(store.lock(key1), 300), timeoutLock(store.lock(key1), 300)])).to.be.rejectedWith(
+        TimeoutError
+      ));
+
+    it('locks by key', () => Promise.all([timeoutLock(store.lock(key1), 300), timeoutLock(store.lock(key2), 300)]));
+
+    it('releases after put', async () => {
+      await store.lock(key1);
+      await expect(timeoutLock(store.lock(key1), 300)).to.be.rejectedWith(TimeoutError);
```

Comment:
Should we check if we can actually get the value between `lock` and `put`?

Reply:
Since we are not using any tokens for locking, anyone can get/put locked key. The point is to prevent code execution while resource is locked. Preventing modifying while locked is not supported by this implementation.